internal/api/installation.go

```diff
@@ -390,6 +395,216 @@ func handleUpdateInstallation(c *Context, w http.ResponseWriter, r *http.Request
 	outputJSON(c, w, installationDTO)
 }
 
+// handleUpdateInstallationVolume responds to POST /api/installation/{installation}/volumes,
+// which adds a custom installation volume.
+func handleCreateInstallationVolume(c *Context, w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	installationID := vars["installation"]
+	c.Logger = c.Logger.WithField("installation", installationID)
+
+	createVolumeRequest, err := model.NewCreateInstallationVolumeRequestFromReader(r.Body)
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to decode request")
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	secretBytes, err := json.Marshal(createVolumeRequest.Data)
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to marshal secret bytes")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	newState := model.InstallationStateUpdateRequested
+
+	installationDTO, status, unlockOnce := getInstallationForTransition(c, installationID, newState)
+	if status != 0 {
+		w.WriteHeader(status)
+		return
+	}
+	defer unlockOnce()
+
+	oldState := installationDTO.State
+
+	err = createVolumeRequest.Apply(installationDTO.Installation)
```

Comment:
Do we need to validate anything with this? Even if as simple as checking for conflicts with current volume names?

Reply:
Yep, we definitely want to do one final validation to look for issues. If you trace out the `Apply` you will end up here were we run some final checks: https://github.com/mattermost/mattermost-cloud/pull/1092/files#diff-888e481ee6df2e28a4dbed85a127b1d24d0a6d7cbefbdda42792d926dc1ec70bR63